### ISSUE TITLE
images: choose a way to resize images

### ISSUE THREAD

**Issue by fabiosantoscode** [iterative/blog#136]
As we know, generating thumbnails and different scales of images, takes a while.

We've already discussed images in [this issue, starting in this comment](https://github.com/iterative/dvc.org/issues/992#issuecomment-589733329), but I would like to outline our choices so we can make an informed decision together.

| Choice | Pros | Cons |
|-----------|--------|------------|
| Cloudflare image resize service | Should be easy to setup | Doesn't work in preview environments |
| Start up our own [imgproxy](https://docs.imgproxy.net/#/) server | Works in preview environments | Multiple buildpacks probably means multiple dynos (?) (the docs aren't clear); Need to sign image URLs before using them, making it harder to implement |
| Resize the images right in the express server | Pretty easy to setup; Avoids arbitrary sizes and images without signing URLs (images are right in the FS; sizes are all in a config file) | Need an async.queue to avoid more than X simultaneous resizes due to CPU (RAM usage is fine because we can [stream files right into sharp and then into the HTTP response](https://github.com/lovell/sharp#stream)) |

All of these will avoid all the resizing business. It's just a matter of picking one, I think. 

I'm a bit biased towards the node server solution. It works locally (gatsby develop and npm start) and in preview environments, should be really easy to set up, and we have a bunch of extra flexibility.

**Comment by fabiosantoscode** [comment]
Requesting comments: @jorgeorpinel @shcheklein @iAdramelk looking forward to the discussion.

**Comment by shcheklein** [comment]
@fabiosantoscode the last option - I suppose Cloudflare will be doing caching?

**Comment by fabiosantoscode** [comment]
@shcheklein indeed, as long as the images have a cache-control header we'll be fine.

**Comment by iAdramelk** [comment]
@fabiosantoscode As I write is another issue, I still prefer to reduce express usage on our side in hopes that Express is a temporary solution. So I'd say that we shouldn't go with the route 3 if possible. 

It will also add us a lot of complexity with async image resizing and possible slowdowns because of them. I observed how `imgproxy` were created from the side and there were a lot of edge cases and problems along the way, so I'm afraid to reimplement it all by ourselves.

I'm ok with options 1 or 2.

P. S. We also didn't really need to implement it right now I think. If we configure correct cache folders on Heroku, we can still have >1m deploys with current amount of images I think. So we can postpone this optimization until amount of our images will significantly grow.

**Comment by fabiosantoscode** [comment]
@iAdramelk I agree that the express way can be problematic here.

The cache on heroku is configured well enough to create fast builds for master, and fast consecutive builds of PR environments. However, the main reason this came up again is that each PR seems to have its own cache on heroku's side (probably for security reasons). So, the first preview environment takes a while to build.

**Comment by fabiosantoscode** [comment]
After having spent some time creating a null-image plugin for remark (so that we can improve it and let it use our resized images, however they may be resized) I found that gatsby-plugin-sharp is still going to want to resize a lot of images due to our templates.

So to create resized images in our non-markdown space we'd need to provide a GraphQL interface to this plugin (which whitelists sizes to avoid abuse).

I think this is a pretty gargantuan task.

## Alternative

To solve the original problem (of having slow builds for the first PR environment build), we can simply sync the public/ folder to S3 after every successful master build (sending only changed files).

Before each build, if for any reason there's no public/ folder we download it from S3.

**Comment by shcheklein** [comment]
@fabiosantoscode our own caching sounds good (S3 or whatnot), but how long will it take us to hit the Heroku limits in terms of the disk size we can use? Are there Heroku plugins  - like shared SSD storage or something?

**Comment by fabiosantoscode** [comment]
@shcheklein The docs were really vague on this point, so I've made a simple experiment to answer your question.

# TL;DR

 - We have ~200mb more to play with right now
 - We have ~10mb before we reach the *soft limit*
 - We can store a large amount of data on-disk during the build if we don't end up deploying it to heroku in the end (so, maybe, deploy somewhere else?)

-------

I've added a `cat /dev/urandom | head -c 1000m > some-file.txt` section to our build script, to see if the build would stop after having an additional gigabyte of space on-disk. It didn't stop because of the extra disk usage, but it failed when deploying to heroku, saying:

```
Compiled slug size: 1.3G is too large (max is 500M).
```

This is because it tried to deploy some-file.txt.

Then I tried again with a 180mb random file. I got this warning:

```
Warning: Your slug size (467 MB) exceeds our soft limit (300 MB) which may affect boot time.
```

**Comment by shcheklein** [comment]
We should probably just put (and sync on build time) all static to S3 and just proxy requests. Any downsides to this approach?

**Comment by shcheklein** [comment]
I'm not sure though how do we build this in the first place if storage is limited by 500Mb

**Comment by fabiosantoscode** [comment]
> We should probably just put (and sync on build time) all static to S3 and just proxy requests. Any downsides to this approach?

Preview environments will require extra work there. Probably we'd need to create a new bucket for each PR manually and fill it up. Luckily there's a way to hook into a preview environment destruction in heroku, where we can collect garbage.

**Comment by fabiosantoscode** [comment]
We no longer need to do this since all static files (includes images) are neatly stored in S3.
